pkg/service/v1alpha1/node.go

```diff
@@ -226,26 +125,47 @@ func (ns *node) NodePublishVolume(
 ) (*csi.NodePublishVolumeResponse, error) {
 
 	var (
-		err        error
-		devicePath string
-		isMounted  bool
+		err           error
+		devicePath    string
+		currentMounts []string
+		vol           *apis.CSIVolume
 	)
 
 	if err = ns.validateNodePublishReq(req); err != nil {
 		return nil, err
 	}
 
+	volumeID := req.GetVolumeId()
+	targetPath := req.GetTargetPath()
 	nodeID := ns.driver.config.NodeID
 
-	vol, err := prepareVolSpecAndWaitForVolumeReady(req, nodeID)
+	err = addVolumeToTransitionList(volumeID, apis.CSIVolumeStatusMountUnderProgress)
 	if err != nil {
-		return nil, status.Error(codes.Internal, err.Error())
+		goto PublishVolumeResponse
 	}
-	if isMounted, err = verifyInprogressAndRecreateCSIVolumeCR(vol); err != nil {
-		return nil, status.Error(codes.Internal, err.Error())
-	} else if isMounted {
-		goto CreateVolumeResponseSuccess
+	defer removeVolumeFromTransitionList(volumeID)
+
+	vol, err = prepareVolSpecAndWaitForVolumeReady(req, nodeID)
+	if err != nil {
+		goto PublishVolumeResponse
```

Comment:
goto even though supported does not seem to be idiomatic

Reply:
goto is helping to avoid a good amount of duplicate code over here.